CARES Act

jdebacker opened the pull request:
This PR seeks to address Issue #2543 by modeling the relevant provisions of the [CARES Act](https://files.taxfoundation.org/20200325223111/FINAL-FINAL-CARES-ACT.pdf).

These include an increase in the ceiling for charitable deductions (`ID_Charity_crt_all`), which was raised to 100% in 2020.  Also, charitable contributions are an above the line deduction (up to $300) for those who take the standard deduction.

hdoupe commented on taxcalc/calcfunctions.py:
```diff
@@ -401,6 +401,49 @@ def AGI(ymod1, c02500, c02900, XTOT, MARS, sep, DSI, exact, nu18, taxable_ubi,
+        number of dependents under age 18
+    taxable_ubi: float
+        taxable UBI amount
+    II_em: 
```
Just an idea that occurred to me in passing: I'm not sure 100% how it would work, but it'd be neat to pull these directly from `policy_current_law.json`. We could probably do this with the records variables, too: https://github.com/PSLmodels/Tax-Calculator/blob/master/taxcalc/records_variables.json

jdebacker replied:
@hdoupe Do you mean the description for the docstring?

hdoupe replied:
Ah, yes that's what I was thinking about.

jdebacker commented:
All tests passing (except for those that check against cached results, which we'd expect to change given changes in law).  New tests were added that test the modified function, `calcfunctions.StdDed()` directly.

[The JCT Estimate](https://www.jct.gov/publications/2020/jcx-12r-20/) on the two individual provisions of the CARES Act that Tax-Calculator is capable of modeling show a revenue loss of about $6.4 billion.

My estimate using Tax-Calculator is just $0.9 billion:
```
# imports
import pandas as pd
import numpy as np
from bokeh.plotting import figure, show
from bokeh.io import output_notebook
import taxcalc

# Create calculators for current law (which now includes CARES_Act) and a repeal of CARES Act
pol = taxcalc.policy.Policy()
rec = taxcalc.records.Records.cps_constructor()
tc_cares = taxcalc.calculator.Calculator(policy=pol, records=rec)
pol.implement_reform({
    'ID_Charity_crt_all': {2020: 0.6, 2026: 0.5},
    'STD_allow_charity_ded_nonitemizers': {2020: False, 2021: False},
    'STD_charity_ded_nonitemizers_max': {2020: 0.0, 2021: 0.0}})
tc_base = taxcalc.calculator.Calculator(policy=pol, records=rec)

# Compute annual revenue for each year in a 10 year budget window
# Do this for with and without the CARES Act
rev_base = {'IIT': {}, 'Payroll': {}}
rev_cares = {'IIT': {}, 'Payroll': {}}
for t in range(2020, 2031):
    tc_base.advance_to_year(t)
    tc_base.calc_all()
    tc_cares.advance_to_year(t)
    tc_cares.calc_all()
    rev_base['IIT'][t] = tc_base.weighted_total('iitax')
    rev_base['Payroll'][t] = tc_base.weighted_total('payrolltax')
    rev_cares['IIT'][t] = tc_cares.weighted_total('iitax')
    rev_cares['Payroll'][t] = tc_cares.weighted_total('payrolltax')
# make table of revenue estimates
rev_base_df = pd.DataFrame.from_dict(rev_base).T
rev_cares_df = pd.DataFrame.from_dict(rev_cares).T
diff_df = (rev_cares_df - rev_base_df) * 1e-9
pd.options.display.float_format = '${:.3f}'.format
diff_df['2021-2030'] = diff_df.sum(axis=1)
diff_df
```

Result:
|         |    2020 |   2021 |   2022 |   2023 |   2024 |   2025 |   2026 |   2027 |   2028 |   2029 |   2030 | 2021-2030 |
|:--------|--------:|-------:|-------:|-------:|-------:|-------:|-------:|-------:|-------:|-------:|-------:|----------:|
| IIT     | $-0.885 | $0.000 | $0.000 | $0.000 | $0.000 | $0.000 | $0.000 | $0.000 | $0.000 | $0.000 | $0.000 |   $-0.885 |
| Payroll |  $0.000 | $0.000 | $0.000 | $0.000 | $0.000 | $0.000 | $0.000 | $0.000 | $0.000 | $0.000 | $0.000 |    $0.000 |

This doesn't seem very close, but I don't know if we should expect such differences due to JCT having better data on charitable contributions.  Or changes in behavior not captured in the extrapolation of underlying data used here (such as high personal income due to economic relief in 2020 and larger than normal percentages of relief going toward charity).

Has Tax-Calculator or taxdata validated against JCT/SOI charitable contribution totals?  Are the differences in revenue estimates expected or acceptable?

cc @MattHJensen @andersonfrailey

MattHJensen commented:
@jdebacker, I imagine there is quite a bit of behavior in JCT's score, and there are also data issues. [This](https://github.com/PSLmodels/taxdata/blob/535a4e4c72c6e80dbcd47055ba8a45d76ca897b0/taxdata/puf/impute_itmex.py) is the relevant section of TaxData for the nonitemizers. These imputations have not been revisited since TCJA or the pandemic as far as I know. I don't know of any imputation of giving above the ceiling for itemizers. 

That said, if we can devise other methods of verifying the logic in this PR, then it seems we could proceed here and then view the difference in scores as issues for Behavioral-Responses, TaxData, and any repos that combine those with Tax-Calculator.

jdebacker commented:
> That said, if we can devise other methods of verifying the logic in this PR, then it seems we could proceed here

I added the above the line deduction for non-itemizers by building off the the existing logic for allowing the deduction of charitable contributions for those taking the standard deduction.  To that, I added the limitation on these deductions (an new parameter) that is set to $300 in the CARES Act.  This only affects the `StdDed`, for which I added a new test, with 9  parameterizations that include `True` and `False` values for `STD_allow_charity_ded_nonitemizers` and, when that is `True`, values of charitable contributions over and under `STD_charity_ded_nonitemizers_max`.

jdebacker commented:
Ah -- just remembered that I used the CPS data in the above.  Looks a lot better with the PUF:

|         |    2020 |   2021 |   2022 |   2023 |   2024 |   2025 |   2026 |   2027 |   2028 |   2029 |   2030 | 2021-2030 |
|:--------|--------:|-------:|-------:|-------:|-------:|-------:|-------:|-------:|-------:|-------:|-------:|----------:|
| IIT     | $-2.535 | $0.000 | $0.000 | $0.000 | $0.000 | $0.000 | $0.000 | $0.000 | $0.000 | $0.000 | $0.000 |   $-2.535 |
| Payroll |  $0.000 | $0.000 | $0.000 | $0.000 | $0.000 | $0.000 | $0.000 | $0.000 | $0.000 | $0.000 | $0.000 |    $0.000 |

MattHJensen commented:
@jdebacker, thanks again for this PR. This LGTM as soon as the cached results are updated and tests are passing.

jdebacker commented:
@MattHJensen Cached results now updated and committed here.  Should be good to go after checks finish.
